src/main/java/org/inferred/freebuilder/processor/excerpt/CheckedBiMap.java

```diff
@@ -0,0 +1,237 @@
+package org.inferred.freebuilder.processor.excerpt;
+
+import com.google.common.collect.BiMap;
+import org.inferred.freebuilder.processor.source.Excerpt;
+import org.inferred.freebuilder.processor.source.LazyName;
+import org.inferred.freebuilder.processor.source.SourceBuilder;
+import org.inferred.freebuilder.processor.source.ValueType;
+
+import java.util.*;
+import java.util.function.BiConsumer;
+
+public class CheckedBiMap extends ValueType implements Excerpt {
+
+  public static final LazyName TYPE = LazyName.of("CheckedBiMap", new CheckedBiMap());
+
+  private CheckedBiMap() {
+  }
+
+  @Override
+  public void addTo(SourceBuilder code) {
+    code.addLine("")
+        .addLine("/**")
+        .addLine(" * A bimap implementation that delegates to a provided put method")
+        .addLine(" * to perform entry validation and insertion into a backing bimap.")
+        .addLine(" */")
+        .addLine("private static class %s<K, V> extends %s<K, V> implements %s<K, V> {",
+            TYPE, AbstractMap.class, BiMap.class
+        )
+        .addLine("")
+        .addLine("  private final %s<K, V> bimap;", BiMap.class)
+        .addLine("  private final %s<K, V> put;", BiConsumer.class)
+        .addLine("")
+        .addLine("  %s(%s<K, V> bimap, %s<K, V> put) {", TYPE, BiMap.class, BiConsumer.class)
+        .addLine("    this.bimap = bimap;")
+        .addLine("    this.put = put;")
+        .addLine("  }")
+        .addLine("")
+        .addLine("  @Override public V get(Object key) {")
+        .addLine("    return bimap.get(key);")
+        .addLine("  }")
+        .addLine("")
+        .addLine("  @Override public boolean containsKey(Object key) {")
+        .addLine("    return bimap.containsKey(key);")
+        .addLine("  }")
+        .addLine("")
+        .addLine("  @Override public V put(K key, V value) {")
+        .addLine("    V oldValue = bimap.get(key);")
+        .addLine("    put.accept(key, value);")
+        .addLine("    return oldValue;")
+        .addLine("  }")
+        .addLine("")
+        .addLine("  @Override public V forcePut(K key, V value) {")
```

Comment:
I'm not quite sure if this would be the correct behavior for forcePut. It was unclear to me how to rationalize forcePut's javadoc with the put BiConsumer.

Reply:
I'd suggest passing around `forcePut` instead of `put`, and doing the IllegalArgumentException-throwing check of `put` explicitly in the put method. You'll also need tests.

Reply:
I've added a forcePut, and defer to forcePut for put, while throwing. It's not clear to me how  to get a BiMap into an illegal state.